src/basic/MAT_FUSED_MUL_ADD-Hip.cpp

```diff
@@ -0,0 +1,304 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "MAT_FUSED_MUL_ADD.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_HIP)
+
+#include "common/HipDataUtils.hpp"
+
+#include <iostream>
+
+namespace rajaperf {
+namespace basic {
+
+#define MAT_FUSED_MUL_ADD_DATA_SETUP_HIP           \
+  const Index_type N = m_N;                        \
+  allocAndInitHipDeviceData(A, m_A, N);            \
+  allocAndInitHipDeviceData(B, m_B, N);            \
+  allocAndInitHipDeviceData(D, m_D, N);			   
+
+#define MAT_FUSED_MUL_ADD_DATA_TEARDOWN_HIP        \
+  getHipDeviceData(m_A, A, N);                     \
+  getHipDeviceData(m_B, B, N);                     \
+  getHipDeviceData(m_D, D, N);                     \
+  deallocHipDeviceData(A);                         \
+  deallocHipDeviceData(B);                         \
+  deallocHipDeviceData(D);						   
+
+__global__ void mat_fused_mul_add_builtin(const Real_ptr A, const Real_ptr B, Real_ptr D, const Index_type N){ 
+  constexpr Index_type Ne = 16;
+  const Index_type N_Elem = N/(Ne*Ne);
+  for(Index_type ii = 0; ii != N_Elem; ++ii){
```

Comment:
Do we want to parallelise over elements?

Reply:
Since this is to mirror an FE mass matrix solve we need to assume that each element is independent and shares no common data (they could but this is the worst case).